### PR TITLE
Removed tempfile dependency

### LINKED ISSUE #1125: Remove tempfile dependency
The `tempfile` crate adds 17 dependencies and it is only used in tests. Instead of using `tempfile` we can create a tiny helper that handles creating files and deleting them when the object gets out of scope.

Functionality needed by Firecracker:
- create directories
- create files

### PR DESCRIPTION
**Pull request by iulianbarbu**
Signed-off-by: Iulian Barbu <iul@amazon.com>

## Reason for This PR

Fixes #1125 

## Description of Changes

Some of Firecracker crates have dev-dependency on `tempfile` crate from crates.io. This crate has other dependencies on its own that are brought into Firecracker as well.

Removed `tempfile` crate dev-dependency and used instead `vmm-sys-util` `tempfile module`, which basically replaces `tempfile` crate with a simpler implementation for temporary files & temporary directories, without additional dependencies.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.
- [x] Either no new `unsafe` code has been added, or, the newly added `unsafe`
      code is unavoidable and properly documented.

### CODE REVIEW

File: devices/src/virtio/block.rs [hunk 1]
```diff
@@ -707,13 +706,23 @@ mod tests {
...
+            tmp_f.as_file().set_len(0x1000).unwrap();
 
             // Rate limiting is enabled but with a high operation rate (10 million ops/s).
             let rate_limiter = RateLimiter::new(0, None, 0, 100_000, None, 10).unwrap();
             DummyBlock {
-                block: Block::new(f, is_disk_read_only, epoll_config, Some(rate_limiter)).unwrap(),
+                block: Block::new(
+                    OpenOptions::new()
+                        .read(true)
+                        .write(true)
```
Comment (andreeaflorescu) [inline devices/src/virtio/block.rs]: Shouldn't we add open options based on `is_disk_read_only` flag?

Comment (iulianbarbu) [reply devices/src/virtio/block.rs]: Yes, I agree. I've updated, please take a look: [devices/src/virtio/block.rs:716](https://github.com/firecracker-microvm/firecracker/blob/145621d1bf024f5a692288b22e762de8ea0b9534/devices/src/virtio/block.rs#L716).

File: devices/src/virtio/block.rs [hunk 2]
```diff
@@ -654,11 +654,10 @@ impl VirtioDevice for Block {
...
-    extern crate tempfile;
+    extern crate utils;
+    use utils::tempfile::TempFile;
 
-    use self::tempfile::{tempfile, NamedTempFile};
     use super::*;
-
     use libc;
     use memory_model::Address;
     use std::fs::{metadata, OpenOptions};
```
Comment (lauralt) [inline devices/src/virtio/block.rs]: I think it would be nice to have uniformity also in tests and keep separate the `extern`s, the `use`s from `std` and the ones from other crates.

File: logger/src/lib.rs [hunk 3]
```diff
@@ -640,12 +640,13 @@ impl Log for Logger {
...
+    extern crate utils;
+    use utils::tempfile::TempFile;
+
     use super::*;
     use log::MetadataBuilder;
 
-    use std::fs::File;
+    use std::fs::{File, OpenOptions};
     use std::io::BufRead;
     use std::io::BufReader;
```
Comment (lauralt) [inline logger/src/lib.rs]: Same comment as above.

File: devices/src/virtio/block.rs [hunk 4]
```diff
@@ -1513,8 +1522,8 @@ mod tests {
 
         // test block device update handler
         {
-            let f = NamedTempFile::new().unwrap();
-            let path = f.path().to_path_buf();
+            let f = TempFile::new("/tmp/").unwrap();
+            let path = f.as_path().to_path_buf();
```
Comment (lauralt) [inline devices/src/virtio/block.rs]: Can't we use here just `let path = f.as_path()`?

File: jailer/src/cgroup.rs [hunk 5]
```diff
@@ -250,35 +252,38 @@ mod tests {
         // 1. If parent file does not exist, return an error.
 
         // This is /A/B/ .
-        let dir = tempdir().expect("Cannot create temporary directory.");
+        let dir = TempDir::new("/tmp/").expect("Cannot create temporary directory.");
         // This is /A/B/C .
-        let dir2 = tempdir_in(dir.path()).expect("Cannot create temporary directory.");
-        let mut path2 = PathBuf::from(dir2.path());
+        println!("{:?}", dir.as_path().join("X"));
```
Comment (lauralt) [inline jailer/src/cgroup.rs]: This `print` is a leftover I guess.

File: devices/src/virtio/block.rs (hunk 2 above)
Comment (iulianbarbu) [reply devices/src/virtio/block.rs]: Updated. Take a look: [devices/src/virtio/block.rs:657](https://github.com/firecracker-microvm/firecracker/blob/f1754fb3c54889c98231bfe3cdcf023c3d74fcf3/devices/src/virtio/block.rs#L657)

File: logger/src/lib.rs (hunk 3 above)
Comment (iulianbarbu) [reply logger/src/lib.rs]: Updated: [logger/src/lib.rs:643](https://github.com/firecracker-microvm/firecracker/blob/f1754fb3c54889c98231bfe3cdcf023c3d74fcf3/logger/src/lib.rs#L643).

File: devices/src/virtio/block.rs (hunk 4 above)
Comment (iulianbarbu) [reply devices/src/virtio/block.rs]: Sure. Done that: [devices/src/virtio/block.rs:1527](https://github.com/firecracker-microvm/firecracker/blob/87231aa9c2462d00b58edf264d1a25482696281a/devices/src/virtio/block.rs#L1528)

File: jailer/src/cgroup.rs (hunk 5 above)
Comment (iulianbarbu) [reply jailer/src/cgroup.rs]: Yes. Removed it. Take a look: [jailer/src/cgroup.rs:257](https://github.com/firecracker-microvm/firecracker/blob/87231aa9c2462d00b58edf264d1a25482696281a/jailer/src/cgroup.rs#L257)

File: devices/src/virtio/block.rs [hunk 6]
```diff
@@ -707,13 +707,23 @@ mod tests {
 
             let epoll_config = EpollConfig::new(0, epoll_raw_fd, sender);
 
-            let f: File = tempfile().unwrap();
-            f.set_len(0x1000).unwrap();
+            let tmp_f = TempFile::new("/tmp/").unwrap();
```
Comment (serban300) [inline devices/src/virtio/block.rs]: "`/tmp`" is very often used. We should probably define it in a constant or have a macro that calls `TempFile::new("/tmp/")`

Comment (iulianbarbu) [reply devices/src/virtio/block.rs]: Created wrappers with the same name, but actually which defaults tmp files creation to "/tmp/". Please take a look: [utils/src/tempfile.rs](https://github.com/firecracker-microvm/firecracker/blob/ca185fa1c1a668a1f69cdb7012b40fd8a32a7216/utils/src/tempdir.rs) and [utils/src/tempdir.rs](https://github.com/firecracker-microvm/firecracker/blob/ca185fa1c1a668a1f69cdb7012b40fd8a32a7216/utils/src/tempdir.rs).

File: devices/src/virtio/block.rs (hunk 2 above)
Comment (lauralt) [reply devices/src/virtio/block.rs]: :thumbsup: Looks great now.

File: utils/src/tempfile.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+use std::path::Path;
+
+use vmm_sys_util::{errno::Result, tempfile};
+
+/// Wrapper over vmm_sys_util::tempfile::TempFile::new().
+/// This creates a temporary file inside $TMPDIR if set,
+/// otherwise inside `/tmp` if non-Android system.
+pub fn tempfile() -> Result<tempfile::TempFile> {
+    let mut in_tmp_dir = temp_dir();
+    in_tmp_dir.push("");
```
Comment (serban300) [inline utils/src/tempfile.rs]: why is this needed ?

File: utils/src/tempfile.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+/// Wrapper over vmm_sys_util::tempfile::TempFile::new().
+/// This creates a temporary directory inside a specific directory,
+/// otherwise inside `/tmp` if non-Android system.
+///
+/// # Arguments
+///
+/// * `path` - A path to a specific directory (e.g "/tmp/x" or "/tmp/x/").
+pub fn tempfile_in(path: &Path) -> Result<tempfile::TempFile> {
+    let mut path_buf = path.canonicalize().unwrap();
+    path_buf.push("");
```
Comment (serban300) [inline utils/src/tempfile.rs]: why is this needed ?

File: utils/src/tempfile.rs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+mod tests {
+
+    use std::ffi::OsStr;
+    use std::path::Component;
+
+    use crate::tempdir::{tempdir, tempdir_in};
+    use crate::tempfile::tempfile;
+
+    #[test]
+    fn test_tempfile_new() {
```
Comment (serban300) [inline utils/src/tempfile.rs]: the method name seems outdated.

File: utils/src/tempfile.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+use std::env::temp_dir;
+use std::path::Path;
+
+use vmm_sys_util::{errno::Result, tempfile};
+
+/// Wrapper over vmm_sys_util::tempfile::TempFile::new().
+/// This creates a temporary file inside $TMPDIR if set,
+/// otherwise inside `/tmp` if non-Android system.
+pub fn tempfile() -> Result<tempfile::TempFile> {
```
Comment (serban300) [inline utils/src/tempfile.rs]: I'm not sure if this should be called simply `tempfile()` or if something like `create_temfile()` would be better.

File: utils/src/tempfile.rs (hunk 7 above)
Comment (iulianbarbu) [reply utils/src/tempfile.rs]: I'll leave a comment with the reason above the line in the code.

The reason is that the call for `temp_dir` is returning a `PathBuf` with the content `/tmp`. When creating a `TempFile::new(prefix)`, the `prefix` needs to be an explicit path/prefix where the temporary file name (a random string of 5 chars) will be placed/appended to.

If we would have just the path `/tmp`, when we will try to create a temporary file inside the `/tmp` directory, that will not happen because, the actual call to `TempFile::new("/tmp")` will create a file which will have a name similiar to this `/tmpXVFGB`, which is incorrect.

The call to `in_tmp_dir.push("")` will append a trailing `/`, which will result in `/tmp/` and which will make successful the call to `TempFile::new(in_tmp_dir)`.

File: utils/src/tempfile.rs (hunk 8 above)
Comment (iulianbarbu) [reply utils/src/tempfile.rs]: Same as above.

File: utils/src/tempfile.rs (hunk 10 above)
Comment (iulianbarbu) [reply utils/src/tempfile.rs]: I'd rather keep it simple. I admit that I am biased because of the API offered by `tempfile` crate from crates.io, which exposes the method with the same name for creating an "anonymous" temporary file. 

The method is documented and commented, so its usage should be pretty clear, while its name, although does not explicitly suggest creation of a tempfile, it surely implies it, because you can see it as a short form of `get_tempfile()`.

Comment (andreeaflorescu) [inline utils/src/tempfile.rs]: I was picturing this more like a new Tempfile structure that extends the existing Tempfile from vmm-sys-util. I also thing that the goal should be to merge this changes in vmm-sys-util. To unblock ourselves it's okay to have these changes in Firecracker.

I quickly typed something that works as an example (it can be made nicer):
```rust
pub struct TempFile(tempfile::TempFile);

impl TempFile {
    fn new() -> Result<Self> {
        let mut in_tmp_dir = temp_dir();
        in_tmp_dir.push("");
        let temp_file = tempfile::TempFile::new(in_tmp_dir)?;
        Ok(TempFile(temp_file))
    }

    fn new_with_path(path: &Path) -> Result<Self> {
        let mut in_tmp_dir = temp_dir();
        let temp_file = tempfile::TempFile::new(in_tmp_dir)?;
        Ok(TempFile(temp_file))
    }

    fn path_ref(&self) -> &Path {
        self.0.as_path()
    }
}
```

The reason for which I prefer a structure is because I don't particularly like orphan functions, but this is subjective. Let me know what you think.

Comment (andreeaflorescu) [reply utils/src/tempfile.rs]: Update: you don't need path_ref as you can implement deref for TempFile.

```
impl Deref for TempFile {
    type Target = tempfile::TempFile;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

Comment (iulianbarbu) [reply utils/src/tempfile.rs]: Sure, this would work better, since it resembles the struct from `vmm-sys-utils`. I'll make it so.

File: utils/src/tempdir.rs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use std::env::temp_dir;
+use std::path::Path;
+
+use vmm_sys_util::{errno::Result, tempdir};
+
+/// Wrapper over vmm_sys_util::tempdir::TempDir::new().
+/// This creates a temporary directory inside $TMPDIR if set,
+/// otherwise inside `/tmp` if non-Android system.
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply utils/src/tempdir.rs]: I'll remove the note, because it is useless in our case. In Android systems the "/tmp" directory is "/data/local/tmp".

File: src/jailer/Cargo.toml [hunk 12]
```diff
@@ -11,4 +11,4 @@ regex = ">=1.0.0"
 utils = { path = "../utils" }
 
 [dev-dependencies]
-tempfile = ">=3.0.2"
+utils = { path = "../utils" }
```
Comment (acatangiu) [inline src/jailer/Cargo.toml]: not required, already present in `dependencies`

File: src/logger/Cargo.toml [hunk 13]
```diff
@@ -14,4 +14,5 @@ serde_json = ">=1.0.9"
 utils = { path = "../utils" }
 
 [dev-dependencies]
-tempfile = ">=3.0.2"
+utils = { path = "../utils" }
```
Comment (acatangiu) [inline src/logger/Cargo.toml]: same here, not required.

File: src/logger/src/lib.rs [hunk 14]
```diff
@@ -640,15 +640,17 @@ impl Log for Logger {
 
 #[cfg(test)]
 mod tests {
-    extern crate tempfile;
-    use self::tempfile::NamedTempFile;
-    use super::*;
-    use log::MetadataBuilder;
+    extern crate utils;
```
Comment (acatangiu) [inline src/logger/src/lib.rs]: Not required, already declared outside of tests.

File: src/memory_model/Cargo.toml [hunk 15]
```diff
@@ -8,4 +8,4 @@ libc = ">=0.2.39"
 utils = { path = "../utils" }
 
 [dev-dependencies]
-tempfile = ">=3.0.2"
+utils = { path = "../utils" }
```
Comment (acatangiu) [inline src/memory_model/Cargo.toml]: not required

File: src/utils/src/tempdir.rs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+impl TempDir {
+    pub fn new() -> Result<Self> {
+        let mut in_tmp_dir = temp_dir();
+        // This `push` adds a trailing slash ("/tmp" -> "/tmp/").
+        // This is safe for paths with already trailing slash.
+        in_tmp_dir.push("");
+        let temp_dir = tempdir::TempDir::new(in_tmp_dir)?;
+        Ok(TempDir(temp_dir))
+    }
```
Comment (acatangiu) [inline src/utils/src/tempdir.rs]: I would've liked to have this directly in `vmm-sys-util` and not wrap it ourselved in Firecracker. I say you open a PR there as well and not let this dangling.

File: src/vmm/Cargo.toml [hunk 17]
```diff
@@ -28,4 +28,5 @@ seccomp = { path = "../seccomp" }
 cpuid = { path = "../cpuid" }
 
 [dev-dependencies]
-tempfile = ">=3.0.2"
+utils = { path = "../utils" }
```
Comment (acatangiu) [inline src/vmm/Cargo.toml]:
```suggestion
```

File: src/vmm/src/lib.rs [hunk 18]
```diff
@@ -1656,19 +1656,18 @@ mod tests {
         }};
     }
 
-    extern crate tempfile;
-
-    use super::*;
+    extern crate utils;
```
Comment (acatangiu) [inline src/vmm/src/lib.rs]:
```suggestion
```

And while you're at it please also remove:
```rust
extern crate logger;
extern crate utils;
```

in `src/vmm/src/signal_handler.rs`.

File: src/vmm/src/vmm_config/drive.rs [hunk 19]
```diff
@@ -228,10 +228,10 @@ impl BlockDeviceConfigs {
 
 #[cfg(test)]
 mod tests {
-    extern crate tempfile;
+    extern crate utils;
```
Comment (acatangiu) [inline src/vmm/src/vmm_config/drive.rs]:
```suggestion
```

File: src/vmm/src/vmm_config/logger.rs [hunk 20]
```diff
@@ -128,15 +128,16 @@ impl Display for LoggerConfigError {
 
 #[cfg(test)]
 mod tests {
-    extern crate tempfile;
-    use self::tempfile::NamedTempFile;
+    extern crate utils;
```
Comment (acatangiu) [inline src/vmm/src/vmm_config/logger.rs]:
```suggestion
```

File: src/utils/src/tempdir.rs (hunk 16 above)
Comment (iulianbarbu) [reply src/utils/src/tempdir.rs]: I agree. We should expose this functionality in `vmm-sys-util`. I will open a PR there soon.

**Comment by iulianbarbu** [comment]
I've included the received feedback. Please, take a look at https://github.com/firecracker-microvm/firecracker/commit/239923316abc5cfea1e8759e5929652287e3656a.

**Comment by dianpopa** [comment]
Will continue work here only after https://github.com/rust-vmm/vmm-sys-util/pull/54 had been merged.

File: src/devices/src/virtio/block.rs [hunk 21]
```diff
@@ -707,13 +707,23 @@ mod tests {
...
-            f.set_len(0x1000).unwrap();
+            let tmp_f = TempFile::new().unwrap();
+            tmp_f.as_file().set_len(0x1000).unwrap();
 
             // Rate limiting is enabled but with a high operation rate (10 million ops/s).
             let rate_limiter = RateLimiter::new(0, None, 0, 100_000, None, 10).unwrap();
             DummyBlock {
-                block: Block::new(f, is_disk_read_only, epoll_config, Some(rate_limiter)).unwrap(),
+                block: Block::new(
+                    OpenOptions::new()
```
Comment (dianpopa) [inline src/devices/src/virtio/block.rs]: Why not `tmp_f.as_file().try_clone().unwrap()` instead of `OpenOptions::new()..`?

File: src/memory_model/src/mmap.rs [hunk 22]
```diff
@@ -383,7 +384,11 @@ mod tests {
             .read_to_memory(1, &mut file, mem::size_of::<u32>())
             .is_ok());
 
-        let mut f = tempfile().unwrap();
+        let temp_f = TempFile::new().unwrap();
```
Comment (dianpopa) [inline src/memory_model/src/mmap.rs]: When TempFile will implement Read, you can do:
```
let mut temp_f = ...
`.read_to_memory(1, &mut temp_f, mem::size_of::<u32>())`
```

File: src/logger/src/lib.rs [hunk 23]
```diff
@@ -729,28 +729,56 @@ mod tests {
...
         let metrics_file_temp =
-            NamedTempFile::new().expect("Failed to create temporary metrics logging file.");
+            TempFile::new().expect("Failed to create temporary metrics logging file.");
         l.set_instance_id(TEST_INSTANCE_ID.to_string());
         assert!(l
             .init(
                 &app_info,
-                Box::new(log_file_temp),
-                Box::new(metrics_file_temp),
+                Box::new(
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/logger/src/lib.rs]: Actually, because it might spread to another week, what do you think if we merge this like it is, if we get the approves? We can update with your suggestion in a subsequent PR. Thinking about it, this PR should've not been blocked at all. The updates could've been added in subsequent PRs (TempFile API changes, signal API refactoring). Please let me know what you think. I just want to not get blocked here everytime we find things that miss or need changes in `rust-vmm/vmm-sys-util`.

File: src/devices/Cargo.toml [hunk 24]
```diff
@@ -19,4 +19,3 @@ virtio_gen = { path = "../virtio_gen" }
 
 [dev-dependencies]
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/devices/Cargo.toml]: `Remove tempfile ...` is inaccurate, I agree, but `Remove tempfile dependency` it has something to do with the dependency, not with the tempfile. I think the commit needs a proper description, to log basically what the change has achieved. I'll add a description.

File: src/logger/src/lib.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (dianpopa) [reply src/logger/src/lib.rs]: I agree

**Comment by acatangiu** [comment]
@iulianbarbu if we merge it like this, please add an issue describing the leftover improvements so we can track them.

**Comment by iulianbarbu** [comment]
> @iulianbarbu if we merge it like this, please add an issue describing the leftover improvements so we can track them.

I've added the following issue: https://github.com/firecracker-microvm/firecracker/issues/1466.
